### PR TITLE
micro: Printed memory plan layout changes

### LINKED ISSUE #48606: Enhance GreedyMemoryPlanner::PrintMemoryPlan() format
@tensorflow/micro

**System information**
- Host OS Platform and Distribution: all
- TensorFlow installed from source
- Tensorflow version: All
- Target platform: All

**Describe the problem**

While using the output of `GreedyMemoryPlanner::PrintMemoryPlan()` to better understand model memory use, I found myself tweaking the output to provide information in a more readily understandable way.

I have a PR that collects these changes: #48595 As noted there, the changes are:

Per-buffer info line improvements:

(a) Reduces text quantity to make it easier to scan for information.
(b) Moves size to near front line, to make it simpler for the eye to
find this information.
(c) Includes ordinal letter used in per-time display below to make it
simpler to cross-reference.

Per time improvements:

(a) Includes tick number. This is useful in larger models for cross
referencing to the per-buffer information and also for helping to
determine the actual operation being executed at that time.
(b) Includes total memory use of buffers, which helps to more clearly
identify memory bottlenecks.

**Please provide the exact sequence of commands/steps when you ran into the problem**

### PR DESCRIPTION
**Pull request by alanvgreen**
Reformats output of GreedyMemoryPlanner:PrintMemoryPlan() to make it
simpler to interpret.

Per-buffer info line improvements:

(a) Reduces text quantity to make it easier to scan for information.
(b) Moves size to near front line, to make it simpler for the eye to
    find this information.
(c) Includes ordinal letter used in per-time display below to make it
    simpler to cross-reference.

Per time improvements:

(a) Includes tick number. This is useful in larger models for cross
    referencing to the per-buffer information and also for helping to
    determine the actual operation being executed at that time.
(b) Includes total memory use of buffers, which helps to more clearly
    identify memory bottlenecks.

### CODE REVIEW

**Comment by alanvgreen** [comment]
Sample output from PrintMemoryPlan()

```
0 (id=0): size=256, offset=2304, first_used=27 last_used=28
1 (id=1): size=16, offset=0, first_used=28 last_used=29
2 (id=2): size=16, offset=16, first_used=29 last_used=30
3 (id=3): size=18432, offset=0, first_used=0 last_used=1
4 (id=4): size=4608, offset=4608, first_used=19 last_used=20
5 (id=5): size=4608, offset=0, first_used=20 last_used=21
6 (id=6): size=4608, offset=4608, first_used=21 last_used=22
7 (id=7): size=4608, offset=0, first_used=22 last_used=23
8 (id=8): size=1152, offset=4608, first_used=23 last_used=24
9 (id=9): size=2304, offset=0, first_used=24 last_used=25
a (id=10): size=2304, offset=2304, first_used=25 last_used=26
b (id=11): size=2304, offset=0, first_used=26 last_used=27
c (id=12): size=18432, offset=36864, first_used=1 last_used=2
d (id=13): size=36864, offset=0, first_used=2 last_used=3
e (id=14): size=9216, offset=36864, first_used=3 last_used=4
f (id=15): size=18432, offset=0, first_used=4 last_used=5
g (id=16): size=18432, offset=18432, first_used=5 last_used=6
h (id=17): size=18432, offset=0, first_used=6 last_used=7
i (id=18): size=4608, offset=18432, first_used=7 last_used=8
j (id=19): size=9216, offset=0, first_used=8 last_used=9
k (id=20): size=9216, offset=9216, first_used=9 last_used=10
l (id=21): size=9216, offset=0, first_used=10 last_used=11
m (id=22): size=2304, offset=9216, first_used=11 last_used=12
n (id=23): size=4608, offset=0, first_used=12 last_used=13
o (id=24): size=4608, offset=4608, first_used=13 last_used=14
p (id=25): size=4608, offset=0, first_used=14 last_used=15
q (id=26): size=4608, offset=4608, first_used=15 last_used=16
r (id=27): size=4608, offset=0, first_used=16 last_used=17
s (id=28): size=4608, offset=4608, first_used=17 last_used=18
t (id=29): size=4608, offset=0, first_used=18 last_used=19
u (id=30): size=16, offset=0, first_used=30 last_used=30
v (id=31): size=9216, offset=18432, first_used=0 last_used=0
   0: 33333333333333333333333333vvvvvvvvvvvvvv........................................ (27k)
   1: 33333333333333333333333333...........................ccccccccccccccccccccccccccc (36k)
   2: dddddddddddddddddddddddddddddddddddddddddddddddddddddccccccccccccccccccccccccccc (54k)
   3: dddddddddddddddddddddddddddddddddddddddddddddddddddddeeeeeeeeeeeee.............. (45k)
   4: ffffffffffffffffffffffffff...........................eeeeeeeeeeeee.............. (27k)
   5: ffffffffffffffffffffffffffggggggggggggggggggggggggggg........................... (36k)
   6: hhhhhhhhhhhhhhhhhhhhhhhhhhggggggggggggggggggggggggggg........................... (36k)
   7: hhhhhhhhhhhhhhhhhhhhhhhhhhiiiiiii............................................... (23k)
   8: jjjjjjjjjjjjj.............iiiiiii............................................... (14k)
   9: jjjjjjjjjjjjjkkkkkkkkkkkkk...................................................... (18k)
  10: lllllllllllllkkkkkkkkkkkkk...................................................... (18k)
  11: lllllllllllllmmm................................................................ (12k)
  12: nnnnnn.......mmm................................................................ (7k)
  13: nnnnnnooooooo................................................................... (9k)
  14: ppppppooooooo................................................................... (9k)
  15: ppppppqqqqqqq................................................................... (9k)
  16: rrrrrrqqqqqqq................................................................... (9k)
  17: rrrrrrsssssss................................................................... (9k)
  18: ttttttsssssss................................................................... (9k)
  19: tttttt4444444................................................................... (9k)
  20: 5555554444444................................................................... (9k)
  21: 5555556666666................................................................... (9k)
  22: 7777776666666................................................................... (9k)
  23: 77777788........................................................................ (6k)
  24: 999...88........................................................................ (4k)
  25: 999aaa.......................................................................... (5k)
  26: bbbaaa.......................................................................... (5k)
  27: bbb............................................................................. (3k)
  28: ................................................................................ (1k)
  29: ................................................................................ (1k)
  30: ................................................................................ (1k)
```

**Comment by alanvgreen** [comment]
Created issue #48606 for discussion.

File: tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc [hunk 1]
```diff
@@ -315,13 +334,14 @@ void GreedyMemoryPlanner::PrintMemoryPlan(ErrorReporter* error_reporter) {
...
+#ifndef TF_LITE_STRIP_ERROR_STRINGS
+    char s[] = {GetOrdinalCharacter(i), 0};
     TF_LITE_REPORT_ERROR(
         error_reporter,
-        "Planner buffer ID: %d, calculated offset: %d, size required: %d, "
-        "first_time_created: %d, "
-        "last_time_used: %d",
-        i, buffer_offsets_[i], requirements_[i].size,
+        "%s (id=%d): size=%d, offset=%d, first_used=%d last_used=%d", s, i,
+        requirements_[i].size, buffer_offsets_[i],
```
Comment (advaitjain) [inline tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc:342]: we are incrementally moving away from TF_LITE_ERROR_REPORTER in favor of using MicroPrintf:
https://github.com/tensorflow/tensorflow/blob/282fa073abc720a3b668c0aebe607c4105c3effe/tensorflow/lite/micro/micro_error_reporter.h#L26

Can you switch this over to
```cc
MicroPrintf("%s (id=%d): size=%d, offset=%d, first_used=%d last_used=%d", s, i, requirements_[i].size, buffer_offsets_[i],requirements_[i].first_time_used, requirements_[i].last_time_used);
```

File: tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc [hunk 2]
```diff
@@ -356,28 +377,21 @@ void GreedyMemoryPlanner::PrintMemoryPlan(ErrorReporter* error_reporter) {
...
         } else {
           line[n] = '!';
         }
       }
     }
     line[kLineWidth] = 0;
-    TF_LITE_REPORT_ERROR(error_reporter, "%s", (const char*)line);
+
+    TF_LITE_REPORT_ERROR(error_reporter, "%s%d: %s (%dk)", t < 10 ? " " : "", t,
+                         (const char*)line, (memory_use + 1023) / 1024);
```
Comment (advaitjain) [inline tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc:394]: use MicroPrintf instead.

and I believe that the error_reporter param can be dropped from PrintMemoryPlan with this change.

**Review by advaitjain — CHANGES_REQUESTED** [review]
sorry for the delay. looks good and small nit requests only.

**Comment by gbaned** [comment]
@alanvgreen  Can you please check @advaitjain's comments and keep us posted ? Thanks!

**Comment by alanvgreen** [comment]
Apologies for the slow response - will have time to address in the next few days.

**Comment by alanvgreen** [comment]
Done. Please Take a Look.

I had been using a #ifndef TF_LITE_STRIP_ERROR_STRINGS to avoid an undefined variable error in the case that MicroPrintf was stripped from the code. To avoid that, I implemented '%c' on MicroPrintf(). This increases debug build code size by several bytes on x86.

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
I have pushed a commit to fix the bazel build. This PR should be ready to go now.

**Review by advaitjain — APPROVED** [review]

**Pull request merged**